nvme.c

```diff
@@ -1927,6 +1927,115 @@ static int get_boot_part_log(int argc, char **argv, struct command *cmd, struct
 	return err;
 }
 
+static int get_phy_rx_eom_log(int argc, char **argv, struct command *cmd,
+		struct plugin *plugin)
+{
+	const char *desc = "Retrieve Physical Interface Receiver Eye Opening\n"
+		"Measurement log for the given device in decoded format\n"
+		"(default), json or binary.";
+	const char *controller = "Target Controller ID.";
+	struct nvme_phy_rx_eom_log *phy_rx_eom_log;
+	size_t phy_rx_eom_log_len;
+	enum nvme_print_flags flags;
+	struct nvme_dev *dev;
+	int err = -1;
+	__u8 lsp_tmp;
+
+	struct config {
+		__u8	lsp;
+		__u16	controller;
+		char	*output_format;
+	};
+
+	struct config cfg = {
+		.lsp		= 0,
+		.controller	= NVME_LOG_LSI_NONE,
+		.output_format	= "normal",
+	};
+
+	OPT_ARGS(opts) = {
+		OPT_BYTE("lsp",          's', &cfg.lsp,           lsp),
+		OPT_SHRT("controller",   'c', &cfg.controller,    controller),
+		OPT_FMT("output-format", 'o', &cfg.output_format, output_format),
+		OPT_END()
+	};
+
+	err = parse_and_open(&dev, argc, argv, desc, opts);
+	if (err)
+		goto ret;
+
+	err = flags = validate_output_format(cfg.output_format);
+	if (err < 0) {
+		nvme_show_error("Invalid output format");
+		goto close_fd;
+	}
+
+	if (cfg.lsp > 127) {
+		nvme_show_error("invalid lsp param: %u", cfg.lsp);
+		err = -1;
+		goto close_fd;
+	} else if ((cfg.lsp & 3) == 3) {
+		nvme_show_error("invalid measurement quality: %u", cfg.lsp & 3);
+		err = -1;
+		goto close_fd;
+	} else if ((cfg.lsp & 12) == 12) {
+		nvme_show_error("invalid action: %u", cfg.lsp & 12);
+		err = -1;
+		goto close_fd;
+	}
+
+	/* Fetching header to calculate total log length */
+	phy_rx_eom_log_len = sizeof(struct nvme_phy_rx_eom_log);
+	phy_rx_eom_log = malloc(phy_rx_eom_log_len);
```

Comment:
Use the `nvme_alloc` helper to ensure that the alignment of the payload is correct and we have a buffer size of multiple of 4k. There are some buggy firmware out there.

```
phy_rx_oem_log = nvme_alloc(phy_rx_log_len);
if (!phy_rx_oem_log)
    return -ENOMEM;
```